quantum/os_detection.c

```diff
@@ -45,41 +45,67 @@ struct setups_data_t setups_data = {
     .cnt_ff = 0,
 };
 
-os_variant_t detected_os = OS_UNSURE;
+volatile os_variant_t detected_os = OS_UNSURE;
+static os_variant_t   reported_os = OS_UNSURE;
+
+void os_detection_init(void) {
+    // simply send the first update, whatever it is
+    process_detected_host_os(detected_os);
+}
+
+void os_detection_task(void) {
+    os_variant_t current_os = detected_os;
+    if (reported_os != current_os) {
+        process_detected_host_os(current_os);
+    }
+}
+
+void process_detected_host_os(os_variant_t os) {
+    reported_os = os;
+    process_detected_host_os_kb(os);
+}
+
+__attribute__((weak)) bool process_detected_host_os_kb(os_variant_t os) {
+    return process_detected_host_os_user(os);
+}
+
+__attribute__((weak)) bool process_detected_host_os_user(os_variant_t os) {
+    return true;
+}
 
 // Some collected sequences of wLength can be found in tests.
 void make_guess(void) {
-    if (setups_data.count < 3) {
-        return;
-    }
-    if (setups_data.cnt_ff >= 2 && setups_data.cnt_04 >= 1) {
-        detected_os = OS_WINDOWS;
-        return;
-    }
-    if (setups_data.count == setups_data.cnt_ff) {
-        // Linux has 3 packets with 0xFF.
-        detected_os = OS_LINUX;
-        return;
-    }
-    if (setups_data.count == 5 && setups_data.last_wlength == 0xFF && setups_data.cnt_ff == 1 && setups_data.cnt_02 == 2) {
-        detected_os = OS_MACOS;
-        return;
-    }
-    if (setups_data.count == 4 && setups_data.cnt_ff == 0 && setups_data.cnt_02 == 2) {
-        // iOS and iPadOS don't have the last 0xFF packet.
-        detected_os = OS_IOS;
-        return;
-    }
-    if (setups_data.cnt_ff == 0 && setups_data.cnt_02 == 3 && setups_data.cnt_04 == 1) {
-        // This is actually PS5.
-        detected_os = OS_LINUX;
-        return;
-    }
-    if (setups_data.cnt_ff >= 1 && setups_data.cnt_02 == 0 && setups_data.cnt_04 == 0) {
-        // This is actually Quest 2 or Nintendo Switch.
-        detected_os = OS_LINUX;
-        return;
+    if (setups_data.count >= 3) {
```

Comment:
Not sure why, but this change (hiding everything behind the count >= 3) is causing it to fail the ps5 test.  reverting it back fixes the failure,

Reply:
Will check. I think I just restructured the code here?

Reply:
well, unit tests show it comes back as 0 (unsure) instead of 1 (linux), for the ps5 unit test.

Reply:
I get it, but given it was previously guarded, the only possibility I see is that the unit test is sending something that, at first, identifies it as Linux, and then another that causes Unsure to be set again.

The thinking here is that there might be events that may retrigger this evaluation and change wLength without reprocessing the USB descriptors (e.g. a KVM switch switching between machines) so if the previous state was something else, this could be applicable.

Another possibility is to just soft-reset the keyboard when the KVM switches, which is probably better.

But I still need to even find out here, if at all, either of those two flows can be triggered. Either way, I can revert this back and not set/report when unsure.

Reply:
If it helps, `make test:os_detection` should be runnable locally.

Reply:
The PS5 test broke because I forgot a `return;` 🤦🏻

Reply:
@drashna so, I found a solution for the initialization and the instability: debouncing. Basically, I've used the same logic we have for key debouncing -- must be stable for X milliseconds before it gets reported.

That way, not only we prevent reporting too soon, but also reporting intermediate results. And it seems it works great.

I've coded the tests for the full sequence of processing, including intermediate results, and the debouncing to ensure the reports are only sent when stable after a given amount of time.

It seems to be the best solution, for me, given we have zero control of how those numbers are processed and/or within which time frame and/or have a way to report when they are effectively done without some deeper modifications and digging up the code for each USB driver, etc.

I've tested a lot on my keyboard and it works well. I've also debugged quite a bit to ensure only one result (the stable one) was being reported.